### PR TITLE
PCHR-2593: Decorate UIB-DatePicker-Popup

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
As part of this PR, the Date Picker component has been removed, and the same functionality has been implemented by decorating `uib-datepicker-popup` directive.

## Screencast
2642d

## Comments
1.6 Commits moved to https://github.com/civicrm/civihr/pull/2174

---

- [x] Tests Pass

### CODE REVIEW

File: org.civicrm.reqangular/build.js [hunk 1]
```diff
@@ -1,81 +1,80 @@
...
-    'common/mocks': '../test/mocks',
-    'common/vendor/perfect-scrollbar': 'common/vendor/perfect-scrollbar.min'
-  },
-  shim: {
-    'common/angular': {
-      exports: 'angular'
+    baseUrl : 'src',
+    out: 'dist/reqangular.min.js',
+    uglify: {
+        no_mangle: true,
```
Comment (AkA84) [inline org.civicrm.reqangular/build.js]: you have 4 spaces indentation throughout the file

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js [hunk 2]
```diff
@@ -1,35 +1,88 @@
-define([], function () {
+/* eslint-env amd */
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: does the decorator need to be under `angular-date`? this module is really something that i would like to get rid of in the future

Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: also, can you please make it a separate decorator? let's not put all the decorations in one basket

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js [hunk 3]
```diff
@@ -1,35 +1,88 @@
...
   'use strict';
 
   /**
    * Decorates the uib-datepicker-popup directive, making its sibling '.input-group-addon'
    * trigger the datepicker
    */
-  return ['$delegate', function ($delegate) {
+  return ['$delegate', '$compile', function ($delegate, $compile) {
     var directive = $delegate[0];
+    var isMobile = document.getElementsByTagName('body')[0].classList.contains('mobile');
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: Could you abstract this in a separate service? Something like `detectDevice`
```js
detectDevice(); // mobile|desktop
```
this way the decorator doesn't need to know how the device is being detected

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js [hunk 4]
```diff
@@ -1,35 +1,88 @@
...
     var origCompile = directive.compile;
+    var calendarIcon = '<span ' +
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: `calendarIconMarkup`

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js [hunk 5]
```diff
@@ -1,35 +1,88 @@
...
+
+        return function (scope, element, attrs) {
+          handleCalendarIconandEvent(scope, element, attrs);
+
+          link.apply(this, arguments);
+        };
+      };
+    }
+
+    function convertToHTML5DatepickerFormat (date) {
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: why are you not documenting any of these functions?

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js [hunk 6]
```diff
@@ -1,35 +1,88 @@
...
+    var isMobile = document.getElementsByTagName('body')[0].classList.contains('mobile');
     var origCompile = directive.compile;
+    var calendarIcon = '<span ' +
+      'class="input-group-addon pointer">' +
+      '<i class="fa fa-calendar"></i>' +
+      '</span>';
 
-    directive.compile = function compile() {
-      var link = origCompile.apply(this, arguments);
+    if (isMobile) {
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: too much stuff going on in this `if`, put it into a separate function

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js [hunk 7]
```diff
@@ -1,35 +1,88 @@
...
+            attrs.$set('max', convertToHTML5DatepickerFormat(value));
+          });
+        };
+      };
+    } else {
+      directive.compile = function compile () {
+        var link = origCompile.apply(this, arguments);
+
+        return function (scope, element, attrs) {
+          handleCalendarIconandEvent(scope, element, attrs);
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: `handleCalendarIconAndEvent`

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js (hunk 5 above)
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: in the name of the function you are not saying what you are converting

Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: also, why is this needed?

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js [hunk 8]
```diff
@@ -1,35 +1,88 @@
...
+          link.apply(this, arguments);
+        };
+      };
+    }
+
+    function convertToHTML5DatepickerFormat (date) {
+      return date ? moment(date).format('Y-MM-DD') : '';
+    }
+
+    function handleCalendarIconandEvent (scope, element, attrs) {
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: the name is too generic, what does it mean "handle"? You have to read the implementation of the function to figure it out

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_leave-request-modal.scss [hunk 9]
```diff
@@ -133,6 +133,13 @@ $row-margin-bottom: 15px;
   margin-top: 7px;
 }
 
+.chr_leave-request-modal__expiry-clear {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_leave-request-modal.scss]: what's this? You are not mentioning this in the PR description

File: org.civicrm.reqangular/src/common/decorators/angular-date/xeditable-addon.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
+/* eslint-env amd */
+
+define([], function () {
+  'use strict';
+
+  /**
+   * Decorates the xeditable directive
```
Comment (AkA84) [inline org.civicrm.reqangular/src/common/decorators/angular-date/xeditable-addon.js]: too generic, say how you are decorating it

File: org.civicrm.reqangular/build.js (hunk 1 above)
Comment (deb1990) [reply org.civicrm.reqangular/build.js]: This file was reverted to previous state. But sure I can fix this.

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js (hunk 2 above)
Comment (deb1990) [reply org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: Moved it outside, as discussed.

Comment (deb1990) [reply org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: Done

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js (hunk 4 above)
Comment (deb1990) [reply org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: Moved the code to `angular.element`, so kept previous name.

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js (hunk 6 above)
Comment (deb1990) [reply org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: Organised the code.

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js (hunk 7 above)
Comment (deb1990) [reply org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: After re organizing the code, it does not exist anymore.

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js (hunk 5 above)
Comment (deb1990) [reply org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: Apologies. Added them now.

Comment (deb1990) [reply org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: Changed to `convertDateToHTML5DatepickerFormat`

Comment (deb1990) [reply org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: Added DOC

File: org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js (hunk 8 above)
Comment (deb1990) [reply org.civicrm.reqangular/src/common/decorators/angular-date/datepicker-addon.js]: After re organizing the code, it does not exist anymore.

File: uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_leave-request-modal.scss (hunk 9 above)
Comment (deb1990) [reply uk.co.compucorp.civicrm.hrleaveandabsences/scss/components/_leave-request-modal.scss]: It was part of the reverted code. It got reverted to its previous state.

File: org.civicrm.reqangular/src/common/decorators/angular-date/xeditable-addon.js (hunk 10 above)
Comment (deb1990) [reply org.civicrm.reqangular/src/common/decorators/angular-date/xeditable-addon.js]: Done

**Comment by deb1990** [comment]
1.6 Changes were moved to https://github.com/civicrm/civihr/pull/2174
Old Datepicker component reverted from 1.7 wip - https://github.com/civicrm/civihr/pull/2175
New Hotfix applied to 1.7-wip- https://github.com/civicrm/civihr/pull/2176 , which has same file changes as this PR. This is done, because otherwise 1.7-wip would have broken datepicker UI.
